presto-hive/src/test/java/com/facebook/presto/hive/util/TestAsyncQueue.java

```diff
@@ -0,0 +1,258 @@
+        queue.offer(5);
+
+        Runnable runnable = () -> {
+            for (int i = 0; i < 700; i++) {
+                getFutureValue(queue.borrowBatchAsync(3, elements -> new BorrowResult<>(elements, null)));
+            }
+        };
+
+        Future<?> future1 = executor.submit(runnable);
+        Future<?> future2 = executor.submit(runnable);
+        Future<?> future3 = executor.submit(runnable);
+        future1.get();
+        future2.get();
+        future3.get();
+
+        queue.finish();
+        assertFalse(queue.isFinished());
+
+        AtomicBoolean done = new AtomicBoolean();
+        executor.submit(() -> {
+            if (!done.get()) {
+                assertFalse(queue.isFinished());
+            }
+        });
+
+        future1 = executor.submit(runnable);
+        future2 = executor.submit(runnable);
+        future3 = executor.submit(runnable);
+        future1.get();
+        future2.get();
+        future3.get();
+        done.set(true);
+
+        assertFalse(queue.isFinished());
+        ArrayList<Integer> list = new ArrayList<>(queue.getBatchAsync(100).get());
+        list.sort(Integer::compare);
+        assertEquals(list, ImmutableList.of(1, 2, 3, 4, 5));
+        assertTrue(queue.isFinished());
+    }
+
+    @Test
+    public void testBorrowThrows()
+            throws Exception
+    {
+        // It doesn't matter the exact behavior when the caller-supplied function to borrow fails.
+        // However, it must not block pending futures.
+
+        AsyncQueue<Integer> queue = new AsyncQueue<>(4, executor);
+        queue.offer(1);
+        queue.offer(2);
+        queue.offer(3);
+        queue.offer(4);
+        queue.offer(5);
+
+        ListenableFuture<?> future1 = queue.offer(6);
+        assertFalse(future1.isDone());
+
+        Runnable runnable = () -> {
+            getFutureValue(queue.borrowBatchAsync(1, elements -> {
+                throw new RuntimeException("test fail");
+            }));
+        };
+
+        try {
+            executor.submit(runnable).get();
+            fail("expected failure");
+        }
+        catch (ExecutionException e) {
+            assertContains(e.getMessage(), "test fail");
+        }
+
+        ListenableFuture<?> future2 = queue.offer(7);
+        assertFalse(future1.isDone());
+        assertFalse(future2.isDone());
+        queue.finish();
+        future1.get();
+        future2.get();
+        assertTrue(queue.offer(8).isDone());
+
+        try {
+            executor.submit(runnable).get();
+            fail("expected failure");
+        }
+        catch (ExecutionException e) {
+            assertContains(e.getMessage(), "test fail");
+        }
+
+        assertTrue(queue.offer(9).isDone());
+
+        assertFalse(queue.isFinished());
+        ArrayList<Integer> list = new ArrayList<>(queue.getBatchAsync(100).get());
+        // 1 and 2 were removed by borrow call; 8 and 9 were never inserted because insertion happened after finish.
```

Comment:
I'd add a separate test for " 8 and 9 were never inserted because insertion happened after finish"

Reply:
This is tested in `testOfferAfterFinish`